ext/http/response_body.rs

```diff
@@ -0,0 +1,247 @@
+#[derive(Debug, Default)]
+pub struct ResponseBytes(ResponseBytesInner, CompletionHandle);
+
+impl ResponseBytes {
+  pub fn initialize(&mut self, inner: ResponseBytesInner) {
+    debug_assert!(matches!(self.0, ResponseBytesInner::Empty));
+    self.0 = inner;
+  }
+
+  pub fn completion_handle(&self) -> CompletionHandle {
+    self.1.clone()
+  }
+
+  fn complete(&mut self, success: bool) -> ResponseBytesInner {
+    if matches!(self.0, ResponseBytesInner::Done) {
+      return ResponseBytesInner::Done;
+    }
+
+    let current = std::mem::replace(&mut self.0, ResponseBytesInner::Done);
+    self.1.complete(success);
+    current
+  }
+}
+
+impl ResponseBytesInner {
+  pub fn len(&self) -> Option<usize> {
+    match self {
+      Self::Done => Some(0),
+      Self::Empty => Some(0),
+      Self::Bytes(bytes) => Some(bytes.len()),
+      Self::Resource(..) => None,
+      Self::V8Stream(..) => None,
+    }
+  }
+}
+
+impl Body for ResponseBytes {
+  type Data = BufView;
+  type Error = AnyError;
+
+  fn poll_frame(
+    mut self: Pin<&mut Self>,
+    cx: &mut std::task::Context<'_>,
+  ) -> std::task::Poll<Option<Result<Frame<Self::Data>, Self::Error>>> {
+    match &mut self.0 {
+      ResponseBytesInner::Done | ResponseBytesInner::Empty => {
+        unreachable!()
+      }
+      ResponseBytesInner::Bytes(..) => {
+        if let ResponseBytesInner::Bytes(data) = self.complete(true) {
+          std::task::Poll::Ready(Some(Ok(Frame::data(data))))
+        } else {
+          unreachable!()
+        }
+      }
+      ResponseBytesInner::Resource(auto_close, stm, ref mut future) => {
+        match future.poll_unpin(cx) {
+          std::task::Poll::Pending => std::task::Poll::Pending,
+          std::task::Poll::Ready(Err(err)) => {
+            std::task::Poll::Ready(Some(Err(err)))
+          }
+          std::task::Poll::Ready(Ok(buf)) => {
+            if buf.is_empty() {
+              if *auto_close {
+                stm.clone().close();
+              }
+              self.complete(true);
+              return std::task::Poll::Ready(None);
+            }
+            // Re-arm the future
+            *future = stm.clone().read(64 * 1024);
+            std::task::Poll::Ready(Some(Ok(Frame::data(buf))))
+          }
+        }
+      }
+      ResponseBytesInner::V8Stream(stm) => match stm.poll_recv(cx) {
+        std::task::Poll::Pending => std::task::Poll::Pending,
+        std::task::Poll::Ready(Some(buf)) => {
+          std::task::Poll::Ready(Some(Ok(Frame::data(buf))))
+        }
+        std::task::Poll::Ready(None) => {
+          self.complete(true);
+          std::task::Poll::Ready(None)
+        }
+      },
+    }
+  }
+
+  fn is_end_stream(&self) -> bool {
+    matches!(self.0, ResponseBytesInner::Done | ResponseBytesInner::Empty)
+  }
+
+  fn size_hint(&self) -> SizeHint {
+    if let Some(size) = self.0.len() {
+      SizeHint::with_exact(size as u64)
```

Comment:
`Resource` has a `size_hint` also.

Reply:
I'll implement this. I need to understand what hyper's requirements are for this.

Reply:
Implemented for both request and response resources.